src/gg.regression.unity.bots/Runtime/Scripts/ActionManager/RGActionManager.cs

```diff
@@ -306,7 +312,16 @@ public static void SimulateKeyState(Key key, bool isPressed)
         {
             if (key != Key.None)
             {
-                KeyboardEventSender.SendKeyEvent(0, key, isPressed ? KeyState.Down : KeyState.Up);
+                // The new Input System does not support generating multiple keyboard events
+                // within one frame for individual keys. So, we generate events for multiple keys
+                // at once whenever sending keyboard events.
+                _keyStates[key] = isPressed ? KeyState.Down : KeyState.Up;
```

Comment:
I'm concerned that we shouldn't this across frames due to real keyboard input potentially corrupting our state.

can we add some code to Clear _keyStates on the start of each new frame  ?

we could probably then just changer the SendKeysInOneEvent method to only accept a Dictionary not a list.

Reply:
I like the idea of changing SendKeysInOneEvent to accept a dictionary. 

I thought about clearing keyStates at the start of each frame, but the problem is that some games could configure the Input System to process events at a different rate (i.e. not every frame) (see the [Update Mode](https://docs.unity3d.com/Packages/com.unity.inputsystem@1.8/manual/Settings.html#update-mode) setting). If the keyStates buffer is cleared too early, then any new key events sent would cause the previous key state changes to get ignored.

Every StateEvent carries the entire state of the keyboard with it, so I don't actually think it's a problem to have multiple events queued for the same key (it will just land on whatever the last value was).

Reply:
So the problem with changing it to a dictionary, and why I think maybe we keep a running buffer of events still to send in KeyboardEventSender....

Is that in my behaviour bot for stars reach, i send multiple key event requests per update call, and in fact send both key up and key down for the same key in that call.

Example, I determine that I want to crouch starting in this update, but i want to keep moving forward.

To avoid 'rolling'. I have to un-press W, hit CTRL, then press W again.  Since this logic is all in a single update from that application space code, having to manage the event logic on updates myself there would be less than ideal.


Thus, my goal is we end up with something in the keyboard event sender that can handle all the sequential key inputs for the current frame, but be smart enough to do it in the minimum number of update calls while still ordered... so example



Events sent by application

Update 1
A - UP
W - UP
CTRL - DOWN
W - DOWN
D - DOWN
R - DOWN

LateUpdate 1
T - DOWN

Update 2
G - DOWN
R - UP


KeyboardEventSender buffering logic

LateUpdate 1
Send Event with (A-UP, W-UP, CTRL-DOWN, D-DOWN, R-DOWN) - May or may not also include T-DOWN depending on script execution ordering...

LateUpdate 2
Send Event with (W-DOWN, G-DOWN, R-UP) - May or may not also include T-DOWN depending on script execution ordering and whether it was already sent in LateUpdate 1

Reply:
but.. typing that.. maybe that also means that update loop specific code does truly belong in the application layer as it would be game specific... thus looping this conversation back to me agreeing that the dictionary argument seems like the best option

Reply:
I can see how the list would be useful in this scenario, but I think the current implementation of SendKeysInOneEvent wouldn't do what is desired? If it encounters a duplicate key entry (or shift) and then starts a new event, that new event would just result in any key state changes that happened up to that point getting discarded, since the call to `DeltaStateEvent.From` will just make a fresh copy of the keyboard state.

Reply:
see 
https://github.com/Regression-Games/RGUnityBots/blob/main/src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Samples/SimpleRandomMovementBehaviour.cs

for context on how it was being used...

it would be awesome if all the logic in there about trying to group events `List<(Key, KeyState)> keyStates = new();` was un-necessary and you could just naively send key up/down requests one by one and they were handled as a single event appropriately by the sender as efficiently as possible

... maybe that's an unreachable pancea .. but i can dream

Reply:
I thought a bit about this and realized I can use the `InputSystem.onAfterUpdate` hook to clear the key state buffer after it's been processed by the input system (commit pushed). Perhaps similar logic could be moved into KeyboardEventSender as well so we can indeed have the desired single key up/down API (buffer key events until the input system completes an update).